src/controlplane/client/client.go

```diff
@@ -1,203 +1,59 @@
 package client
 
 import (
-	"context"
-	"crypto/tls"
-	"crypto/x509"
 	"fmt"
 	"net/http"
-	"net/url"
-	"path"
-	"strings"
-	"time"
 
 	"github.com/newrelic/infra-integrations-sdk/log"
-	"github.com/pkg/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/rest"
-	"k8s.io/client-go/transport"
 
-	"github.com/newrelic/nri-kubernetes/v2/internal/config"
 	"github.com/newrelic/nri-kubernetes/v2/src/client"
 	"github.com/newrelic/nri-kubernetes/v2/src/prometheus"
 )
 
-type invalidTLSConfig struct {
-	message string
-}
-
-func (i invalidTLSConfig) Error() string {
-	return i.message
-}
-
 type Client struct {
-	httpClient      *http.Client
-	logger          log.Logger
-	k8sClient       kubernetes.Interface
-	InClusterConfig *rest.Config
-	endpoint        *url.URL
-	auth            *config.Auth
+	logger           log.Logger
+	connectionParams *connParams
 }
 
 type Config struct {
-	Logger          log.Logger
-	K8sClient       kubernetes.Interface
-	InClusterConfig *rest.Config
-	EndpoinURL      string
-	Auth            *config.Auth
+	Logger    log.Logger
+	Connector Connector
 }
 
 func New(cfg Config) (client.HTTPClient, error) {
-	if cfg.EndpoinURL == "" {
-		return nil, fmt.Errorf("URL must not be empty")
+	if cfg.Connector == nil {
+		return nil, fmt.Errorf("connector must not be nil")
 	}
 
 	if cfg.Logger == nil {
 		return nil, fmt.Errorf("logger must not be nil")
 	}
 
-	if cfg.InClusterConfig == nil {
-		return nil, fmt.Errorf("InClusterConfig must not be nil")
-	}
-
-	if cfg.Auth != nil && cfg.K8sClient == nil {
-		return nil, fmt.Errorf("k8s client must not be nil if Auth is set")
-	}
-
-	u, err := url.Parse(cfg.EndpoinURL)
+	cp, err := cfg.Connector.Connect()
 	if err != nil {
-		return nil, fmt.Errorf("parsing endpoint url %s: %w", cfg.EndpoinURL, err)
+		return nil, err
 	}
 
 	c := &Client{
-		httpClient:      &http.Client{Timeout: 500 * time.Millisecond},
-		logger:          cfg.Logger,
-		k8sClient:       cfg.K8sClient,
-		InClusterConfig: cfg.InClusterConfig,
-		auth:            cfg.Auth,
-		endpoint:        u,
-	}
-
-	if err = c.configureAuthentication(); err != nil {
-		return nil, fmt.Errorf("fail configuring auth method: %w", err)
+		logger:           cfg.Logger,
+		connectionParams: cp,
 	}
 
 	return c, nil
 }
 
-// Get implements HTTPGetter interface by selecting proper authentication strategy for request
-// based on client configuration.
-//
-// TODO If secure request fails and insecure fallback is configured, request will be attempted over HTTP.
 func (c *Client) Get(urlPath string) (*http.Response, error) {
```

Comment:
urlPath is not used in this function, is there a bug?

Reply:
The others prometheus clients have the method:
```
func (c *Client) MetricFamiliesGetFunc(url string) prometheus.FetchAndFilterMetricsFamilies {
```

implementing: MetricFamiliesGetFunc interface, could we follow the same approach?

Reply:
or is it waiting for the grouper refactor?